tests: logging: syst: Add testcases for mipi syst.

aasthagr opened the pull request:
Adding testcases for mipi syst logging format.

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

nashif commented on west.yml:
```diff
@@ -179,7 +179,8 @@ manifest:
         - debug
-      revision: d9da086b11cda494d85f4d8a9829f505c2d5e380
+      url: https://github.com/aasthagr/mipi-sys-t
+      revision: syst_fix_add_wint_type
```
not the right way to do this, please refer to the docs for how this should be done.

https://docs.zephyrproject.org/latest/guides/modules.html#process-for-submitting-changes-to-existing-modules

aasthagr replied:
Sure. Updated this as per the docs.

laurenmurphyx64 commented:
Hey, I know we were having trouble figuring out how to get the string constants on one line without getting newlines inserted, but I was taking a look at a PR (#43984) Tom just put up and had the idea to do this:

```
#define PAYLOAD "4465627567206D657373616765206578616D706C652C2025642C2025642C202564000"\
	"10000000200000003000000"
const char *payload = PAYLOAD;
...
void test_main(void)
{
	printk("PAYLOAD: %s", PAYLOAD);
...
```

Which printed this:

```
Booting from ROM..*** Booting Zephyr OS build zephyr-v3.0.0-1284-g41805a779e05  ***

PAYLOAD 4465627567206D657373616765206578616D706C652C2025642C2025642C20256400010000000200000003000000

Running test suite test_log_syst...
```

And when I ran the compliance check CI does (see `.github/workflows/compliance.yml:72`), I got no warnings:

```
Running checkpatch       tests in /home/herbert/zephyr-external/zephyr ...
Running Kconfig          tests in /home/herbert/zephyr-external/zephyr ...
Running Codeowners       tests in /home/herbert/zephyr-external/zephyr ...
Running Nits             tests in /home/herbert/zephyr-external/zephyr ...
Running Gitlint          tests in /home/herbert/zephyr-external/zephyr ...
Running pylint           tests in /home/herbert/zephyr-external/zephyr ...
Running Identity         tests in /home/herbert/zephyr-external/zephyr ...

Complete results in compliance.xml
```

Examining the string in the debugger yields one with no breaks or extra characters (copied from Visual Studio Code debugger):

```
payload: 0x1054ac "4465627567206D657373616765206578616D706C652C2025642C2025642C20256400010000000200000003000000"
```

Hopefully this should make it so you don't have to build the string yourself.

aasthagr commented:
> Hey, I know we were having trouble figuring out how to get the string constants in under the line length limit without getting newlines inserted, but I was taking a look at a PR (#43984) Tom just put up and had the idea to do this:
> 
> ```
> #define PAYLOAD "4465627567206D657373616765206578616D706C652C2025642C2025642C202564000"\
> 	"10000000200000003000000"
> const char *payload = PAYLOAD;
> ...
> void test_main(void)
> {
> 	printk("PAYLOAD: %s", payload);
> ...
> ```
> 
> Which printed this:
> 
> ```
> Booting from ROM..*** Booting Zephyr OS build zephyr-v3.0.0-1284-g41805a779e05  ***
> 
> PAYLOAD 4465627567206D657373616765206578616D706C652C2025642C2025642C20256400010000000200000003000000
> 
> Running test suite test_log_syst...
> ```
> 
> And when I ran the compliance check CI does (see `.github/workflows/compliance.yml:72`), I got no warnings:
> 
> ```
> Running checkpatch       tests in /home/herbert/zephyr-external/zephyr ...
> Running Kconfig          tests in /home/herbert/zephyr-external/zephyr ...
> Running Codeowners       tests in /home/herbert/zephyr-external/zephyr ...
> Running Nits             tests in /home/herbert/zephyr-external/zephyr ...
> Running Gitlint          tests in /home/herbert/zephyr-external/zephyr ...
> Running pylint           tests in /home/herbert/zephyr-external/zephyr ...
> Running Identity         tests in /home/herbert/zephyr-external/zephyr ...
> 
> Complete results in compliance.xml
> ```
> 
> Examining the string in the debugger yields one with no breaks or extra characters (copied from Visual Studio Code debugger):
> 
> ```
> payload: 0x1054ac "4465627567206D657373616765206578616D706C652C2025642C2025642C20256400010000000200000003000000"
> ```
> 
> Hopefully this should make it so you don't have to build the string yourself.

@laurenmurphyx64 I had already tried this but checkpatch did complaint atleast in my local environment. May be CI does not complain about it and checkpatch in local still does? I can try though. Thanks!!

aasthagr committed:
tests: logging: syst: Add testcases for mipi syst.

Adding testcases for mipi syst logging format. Excluding
problem architectures from the testcase as SYS-T does not
support 64-bit or big endian architectures. The posix arch
is excluded as it generates numerous anomalous runtime messages
because it does not have the means to determine when data
resides in the rodata section.

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

nashif commented on subsys/logging/log_output_syst.c:
```diff
@@ -308,7 +308,7 @@ static void write_d32(struct mipi_syst_handle *systh, uint32_t v)
 }
 
 #if defined(MIPI_SYST_PCFG_ENABLE_64BIT_IO)
-static void write_d64(struct mipi_syst_handle *systh, uint64_t v)
```
this should be in its own commit explaining why you are changing this.

aasthagr replied:
Not that there is a filter in place for 64 bit architectures, this is not needed. Hence changed it back to uint64_t.

nashif commented on subsys/testsuite/Kconfig:
```diff
@@ -160,4 +160,17 @@ config TEST_BUSY_SIM
 	  It simulates cpu load by using counter device to generate interrupts
 	  with random intervals and random busy looping in the interrupt.
 
+config LOG_BACKEND_MOCK
```
why are you defining this to be part of the testsuite and without any associated code? How can this be used? Maybe put this as part of the test itself if you need that or have it in its own commit and  give some details how this can be used.

aasthagr replied:
Have moved this option to Kconfig under the testcase.

nashif commented on tests/subsys/logging/log_syst/src/main.c:
```diff
@@ -0,0 +1,158 @@
+
+	strcpy(payload, "");
+	strcat(payload, payload_body);
+	strcat(payload, payload_params);
```
I am not sure exactly why are you doing here and why you need all this instead of defining the string directly...

aasthagr replied:
Have removed this. Was using to address the checkpatch warning.

nashif commented on tests/subsys/logging/log_syst/src/main.c:
```diff
@@ -0,0 +1,158 @@
+	"4465627567206D657373616765206578616D706C652C2025642C2025642C202564000";
+	const char *payload_params =
+	"10000000200000003000000";
+	char *payload = malloc(strlen(payload_body)+strlen(payload_params)+1);
```
why is this needed?

aasthagr replied:
Removed this. Was doing this earlier to address the checkpatch warnings treated as error for long string. Have now used the macro instead.

nashif commented on tests/subsys/logging/log_syst/testcase.yaml:
```diff
@@ -0,0 +1,24 @@
+tests:
+  logging.log_syst.mipi_syst:
+    tags: log_syst logging
+    arch_exclude: mips nios2 posix sparc
```
explain why those are excluded

edersondisouza commented on tests/subsys/logging/log_syst/src/main.c:
```diff
@@ -0,0 +1,158 @@
+
+	zassert_equal_ptr(test_log_output_func, log_output_msg2_syst_process,
+					"Correct Function pointer for SYST log\n"
+					"log format was not selected %x vs %x",
```
Why not `%p` to show the pointer value?

aasthagr replied:
yeahh %p is probably better to use here. Changed to %p. Thanks!!

aasthagr committed:
scripts/check_complaince: Add configs to UNDEF_KCONFIG_WHITELIST

Adding LOG_BACKEND_MOCK_OUTPUT_DEFAULT &
LOG_BACKEND_MOCK_OUTPUT_SYST to UNDEF_KCONFIG_WHITELIST as Kconfig
options generated by a template under testcases are not
identified by Compliance check by default.

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

ceolin commented on tests/subsys/logging/log_syst/prj.conf:
```diff
@@ -0,0 +1,21 @@
+CONFIG_ARCH_LOG_LEVEL_OFF=y
+CONFIG_KERNEL_LOG_LEVEL_OFF=y
+CONFIG_LOG_RUNTIME_FILTERING=n
+CONFIG_MAIN_THREAD_PRIORITY=5
```
Do you really need to set the main thread priority ?

aasthagr replied:
Removed the CONFIG_MAIN_THREAD_PRIORITY for now.

ceolin commented on tests/subsys/logging/log_syst/prj.conf:
```diff
@@ -0,0 +1,21 @@
+CONFIG_LOG_RUNTIME_FILTERING=n
+CONFIG_MAIN_THREAD_PRIORITY=5
+CONFIG_LOG_FUNC_NAME_PREFIX_DBG=n
+
```
Same thing, do you need to turn off function name prefix for this test ?

aasthagr replied:
The CONFIG_LOG_FUNC_NAME_PREFIX_DBG is needed as it results in modification of output str which is used for validation here.

ceolin commented on tests/subsys/logging/log_syst/src/mock_backend.c:
```diff
@@ -0,0 +1,136 @@
+static int char_out(uint8_t *data, size_t length, void *ctx)
+{
+	ARG_UNUSED(ctx);
+	size_t output_length = length;
```
you don't need this variable. You can just return length.

aasthagr replied:
length is getting altered in char_out routine. output_length is to return the number of characters copied to the buffer.

ceolin commented on tests/subsys/logging/log_syst/src/mock_backend.c:
```diff
@@ -0,0 +1,136 @@
+		length--;
+
+		if (i == sizeof(test_output_buf)-1) {
+			i = 0;
```
Is it the behavior you want ? Overwrite the buffer data is bigger than the space you have.

aasthagr replied:
When the buffer is full, the buffer size needs to be increased. In our case this block would not have executed since we know the input strings passed from the testcase. But good point, replaced this logic with zassert statement which would ask to increase the size of buffer when it is full instead of overriding it.

ceolin commented on tests/subsys/logging/log_syst/src/mock_backend.c:
```diff
@@ -0,0 +1,136 @@
+
+static uint8_t mock_output_buf[1];
+uint8_t test_output_buf[256];
+int i;
```
That is a bad name for a variable with this scope. Please, rename it.

aasthagr replied:
Renamed. Thanks!!

ceolin commented on tests/subsys/logging/log_syst/src/mock_backend.c:
```diff
@@ -0,0 +1,136 @@
+	const char *syst_format_headers[4] = {type, optional_flags, module_id, sub_type};
+	const char *syst_headers_name[4] = {"type", "optional_flags", "module_id", "sub_type"};
+
+    /* Validate "SYS-T RAW DATA: " prefix in the output_str */
```
wrong identation.

aasthagr replied:
Fixed.

ceolin commented on tests/subsys/logging/log_syst/src/mock_backend.c:
```diff
@@ -0,0 +1,136 @@
+	}
+
+	zassert_mem_equal(output_str, payload, strlen(payload),
+								  "Incorrect Comparison of payload");
```
ditto

aasthagr replied:
Fixed

ceolin commented on tests/subsys/logging/log_syst/src/mock_backend.c:
```diff
@@ -0,0 +1,136 @@
+		}
+	}
+
+	zassert_mem_equal(output_str, payload, strlen(payload),
```
does the payload always has a null terminator ? If not, you can't use `strlen()`

aasthagr replied:
Since the strings are created/defined in main.c from where the validate_msg routine is being called , it will always implicitly have a null terminator.

ceolin commented on tests/subsys/logging/log_syst/src/mock_backend.c:
```diff
@@ -0,0 +1,136 @@
+static void panic(struct log_backend const *const backend)
+{
+
+}
```
These two functions are not doing anything. Can't you just set `.init` and `.panic` to NULL ?

aasthagr replied:
On Some platforms like native_posix, panic is called at the end of test to flush all outputs and if it finds NULL pointer to panic then that results in a segmentation fault. So it is necessary to define these functions even though they are empty.

ceolin commented on tests/subsys/logging/log_syst/src/mock_backend.h:
```diff
@@ -0,0 +1,50 @@
+
+struct mock_log_backend_msg {
+	/* Not used. */
+	uint8_t data[32];
```
What does this comment means ? If that is not used, what is used in `mock_log_backend` ? just `bool panic;` ?

aasthagr replied:
Removed the mock_log_backend_msg and its reference from mock_log_backend struct for now as mock backend is not holding any messages for now. Can add back later if its use is identified.

ceolin commented on tests/subsys/logging/log_syst/src/mock_backend.h:
```diff
@@ -0,0 +1,50 @@
+ */
+void validate_msg(const char *type, const char *optional_flags,
+					const char *module_id, const char *sub_type,
+					const char *payload);
```
fix identation here as well.

aasthagr replied:
Fixed.

edersondisouza commented on tests/subsys/logging/log_syst/src/mock_backend.c:
```diff
@@ -0,0 +1,135 @@
+
+bool isKthBitSet(int n, int k)
+{
+	if (n & (1 << (k - 1)))
```
Nitpick: this could simply be `return (n & (1 << (k - 1)))`

aasthagr replied:
yeahh, totally makes sense :) Changed.

edersondisouza commented on tests/subsys/logging/log_syst/src/mock_backend.c:
```diff
@@ -0,0 +1,135 @@
+
+void validate_msg(const char *type, const char *optional_flags,
+					const char *module_id, const char *sub_type,
+					const char *payload)
```
I think that on function declaration, the goal is to indent the next line at the same column just after the opening parenthesis. Something like:
```
void validate_msg(const char *type, const char *optional_flags,
		  const char *module_id, const char *sub_type,
		  const char *payload)
```
Use tabs to get closer, and spaces to finish indenting.

aasthagr replied:
Done.

edersondisouza commented on tests/subsys/logging/log_syst/src/mock_backend.c:
```diff
@@ -0,0 +1,135 @@
+	 * Optional flags are contained in 1st byte of data in output_str. There are 4 bits
+	 * reserved for Optional flags. 1st bit = Location, 2nd Bit = payload length
+	 * 3rd bit = Sys-t Checksum, 4th bit = Sys-t Timestamp. Support for validating the
+	 * content based on optional flags is not supported, hence the corrosponding
```
s/corrosponding/corresponding/

aasthagr replied:
Fixed

edersondisouza commented on tests/subsys/logging/log_syst/src/mock_backend.c:
```diff
@@ -0,0 +1,135 @@
+	 */
+	size_t skip_bytes[5] = {0, 18, 4, 8, 16};
+
+	for (int i = 1; i < ARRAY_SIZE(skip_bytes); i++) {
```
Why start on one? What about making `skip_bytes` have 4 items only?

aasthagr replied:
Coz optional flags are contained in 1st byte of data. So when we do  _if (isKthBitSet(strtol(optional_flags, 0, 16), i))_ , i needs to start from 1. Either the logic can be tweaked to do i+1 here if skip_bytes contained 4 items or the way it is done works fine.

edersondisouza commented on scripts/ci/check_compliance.py:
```diff
@@ -561,6 +561,8 @@ def get_defined_syms(kconf):
     "REG2",
```
On commit message: s/complaince/compliance/g

aasthagr replied:
Fixed

aasthagr commented on tests/subsys/logging/log_syst/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
```
Fixed

aasthagr commented on tests/subsys/logging/log_syst/Kconfig:
```diff
@@ -0,0 +1,18 @@
+config LOG_BACKEND_MOCK
+	bool "MOCK backend"
+	help
+	  When enabled backend is using MOCK backend to output logs.
```
Modified this. Please re-review.

peter-mitsis commented on tests/subsys/logging/log_syst/src/main.c:
```diff
@@ -0,0 +1,151 @@
+	const char *optional_flags = "0A";
+	const char *module_id = "00";
+	const char *sub_type;
+#ifdef CONFIG_64BIT
+	sub_type = "0C";
+#else
+	sub_type = "0B";
+#endif
```
I note that there are a few places where this logic is duplicated. It may be worthwhile to do something like ...

#ifdef CONFIG_64BIT
    #define SUB_TYPE  "0C"
#else
    #define SUB_TYPE "0B"
#endif

And then in the test code ...
    const char *sub_type = SUB_TYPE;

This would have the advantage of having the conditional logic in one place only.

aasthagr replied:
Makes sense. Have moved additional declarations which were duplicated outside of test functions.

peter-mitsis commented on tests/subsys/logging/log_syst/src/mock_backend.c:
```diff
@@ -0,0 +1,132 @@
+
+bool isKthBitSet(int n, int k)
+{
+	return (n & (1 << (k - 1)));
```
Very minor quibble. If returning a bool, this should be using something like ...

    return ((n & (1 << (k - 1))) != 0);

... otherwise it is returning an integer.

aasthagr replied:
Good catch. Thanks. Addressed!!

peter-mitsis commented on tests/subsys/logging/log_syst/src/main.c:
```diff
@@ -0,0 +1,151 @@
+#else
+	sub_type = "0B";
+#endif
+	const char *payload =
+	"4465627567206D657373616765206578616D706C652C2025640001000000";
```
This comment is a maintenance suggestion. Feel free to ignore.

I note that the payload in each of the tests is comprised of two hex-string parts--format string and arguments. As the hex-strings are somewhat harder to interpret, it may be worthwhile to have an intermediary validation routine that converts a string to a hex-string. Not only should this make it easier to follow what each payload test is verifying, it also provides some flexibility should the format strings change in the future.

nordic-krch commented on tests/subsys/logging/log_syst/src/main.c:
```diff
@@ -0,0 +1,138 @@
+#include <stdlib.h>
+
+#define TEST_MSG_0 "0 args"
+#define PAYLOAD_MULTIPLE_ARGS "4465627567206D657373616765206578616D706C652C2025642C2025642C20256"\
```
any chance of having a comment that would somehow explain this stream?
If it's STP stream maybe something like
`0x4 (D8) 0x46 0x5 (D16) ...` (not sure if i got it correct).

aasthagr replied:
@nordic-krch I have added the comment for PAYLOAD_MULTIPLE_ARGS explaining its purpose.
